// marked/index.js
function inlineText(text) {
  return [{ type: 'text', raw: text, text }];
}

function lex(markdown = '') {
  const text = String(markdown);
  if (!text) return [];

  const tokens = [];
  const htmlCommentPattern = /<!--[\s\S]*?-->/g;
  let lastIndex = 0;
  let match;

  while ((match = htmlCommentPattern.exec(text)) !== null) {
    if (match.index > lastIndex) {
      const raw = text.slice(lastIndex, match.index);
      tokens.push({
        type: 'paragraph',
        raw,
        text: raw,
        tokens: inlineText(raw),
      });
    }
    tokens.push({ type: 'html', raw: match[0], text: match[0] });
    lastIndex = match.index + match[0].length;
  }

  if (lastIndex < text.length) {
    const raw = text.slice(lastIndex);
    tokens.push({
      type: 'paragraph',
      raw,
      text: raw,
      tokens: inlineText(raw),
    });
  }

  return tokens;
}

export class Lexer {
  constructor(_options = {}) {}

  lex(markdown = '') {
    return lex(markdown);
  }
}

export const marked = {
  lexer: lex,
  use() {},
};

marked.Lexer = Lexer;
export default marked;
