// @modelcontextprotocol/sdk/types.js
export const ErrorCode = {
  ParseError: -32700,
  InvalidRequest: -32600,
  MethodNotFound: -32601,
  InvalidParams: -32602,
  InternalError: -32603,
}

export class McpError extends Error {
  constructor(code, message, data) {
    super(message)
    this.code = code
    this.data = data
  }
}

function passthroughSchema() {
  return {
    parse(value) {
      return value
    },
    safeParse(value) {
      return { success: true, data: value }
    },
  }
}

export const CallToolResultSchema = passthroughSchema()
export const CallToolRequestSchema = passthroughSchema()
export const ElicitRequestSchema = passthroughSchema()
export const ElicitationCompleteNotificationSchema = passthroughSchema()
export const JSONRPCMessageSchema = passthroughSchema()
export const ListPromptsResultSchema = passthroughSchema()
export const ListResourcesResultSchema = passthroughSchema()
export const ListRootsRequestSchema = passthroughSchema()
export const ListToolsRequestSchema = passthroughSchema()
export const ListToolsResultSchema = passthroughSchema()
export const PromptListChangedNotificationSchema = passthroughSchema()
export const ReadResourceResultSchema = passthroughSchema()
export const ResourceListChangedNotificationSchema = passthroughSchema()
export const ToolListChangedNotificationSchema = passthroughSchema()
