// @opentelemetry/api/index.js
function spanContext() {
  return {
    traceId: '00000000000000000000000000000000',
    spanId: '0000000000000000',
    traceFlags: 0,
  };
}

function createSpan() {
  return {
    spanContext,
    setAttribute() { return this; },
    setAttributes() { return this; },
    addEvent() { return this; },
    recordException() { return this; },
    setStatus() { return this; },
    updateName() { return this; },
    end() {},
    isRecording() { return false; },
  };
}

const tracer = {
  startSpan() {
    return createSpan();
  },
  startActiveSpan(_name, arg2, arg3, arg4) {
    const span = createSpan();
    const callback = [arg2, arg3, arg4].find(value => typeof value === 'function');
    return callback ? callback(span) : span;
  },
};

export const trace = {
  getTracer() {
    return tracer;
  },
  getActiveSpan() {
    return undefined;
  },
  setSpan(ctx) {
    return ctx;
  },
  setGlobalTracerProvider() {},
};

export const context = {
  active() {
    return {};
  },
  with(_ctx, fn, thisArg, ...args) {
    return fn.apply(thisArg, args);
  },
};

export const diag = {
  setLogger() {},
  disable() {},
  error() {},
  warn() {},
  info() {},
  debug() {},
  verbose() {},
};

export const DiagLogLevel = {
  NONE: 0,
  ERROR: 30,
  WARN: 50,
  INFO: 60,
  DEBUG: 70,
  VERBOSE: 80,
  ALL: 9999,
};

export const SpanStatusCode = {
  UNSET: 0,
  OK: 1,
  ERROR: 2,
};
