// @alcalzone/ansi-tokenize/index.js
const ANSI_RE = /\x1b\[[0-?]*[ -/]*[@-~]|\x1b\][\s\S]*?(?:\x07|\x1b\\)/g;

function endCodeFor(code) {
  if (code === '\x1b[0m') return code;
  if (code === '\x1b[7m') return '\x1b[27m';
  if (code.startsWith('\x1b]8;')) return '\x1b]8;;\x1b\\';
  return '\x1b[0m';
}

export function tokenize(input = '') {
  const text = String(input);
  const tokens = [];
  let lastIndex = 0;
  let match;

  while ((match = ANSI_RE.exec(text)) !== null) {
    if (match.index > lastIndex) {
      tokens.push({
        type: 'text',
        value: text.slice(lastIndex, match.index),
        fullWidth: false,
      });
    }
    tokens.push({
      type: 'ansi',
      code: match[0],
      value: match[0],
      endCode: endCodeFor(match[0]),
    });
    lastIndex = match.index + match[0].length;
  }

  if (lastIndex < text.length) {
    tokens.push({
      type: 'text',
      value: text.slice(lastIndex),
      fullWidth: false,
    });
  }

  return tokens;
}

export function reduceAnsiCodes(codes = []) {
  const result = [];
  for (const code of codes) {
    if (!code) continue;
    if (code.code === '\x1b[0m' || code.code === code.endCode) {
      result.length = 0;
      continue;
    }
    result.push(code);
  }
  return result;
}

export function undoAnsiCodes(codes = []) {
  return [...codes]
    .reverse()
    .map(code => ({ ...code, code: code.endCode, endCode: code.endCode }));
}

export function diffAnsiCodes(_from = [], to = []) {
  const reducedTo = reduceAnsiCodes(to);
  if (reducedTo.length === 0) {
    return _from.length > 0 ? [{ code: '\x1b[0m', endCode: '\x1b[0m' }] : [];
  }
  return [{ code: '\x1b[0m', endCode: '\x1b[0m' }, ...reducedTo];
}

export function ansiCodesToString(codes = []) {
  return codes.map(code => code.code || '').join('');
}

export function styledCharsFromTokens(tokens = []) {
  const chars = [];
  let styles = [];
  for (const token of tokens) {
    if (token.type === 'ansi') {
      styles = reduceAnsiCodes([...styles, token]);
      continue;
    }
    for (const value of [...(token.value || '')]) {
      chars.push({ value, styles });
    }
  }
  return chars;
}
